File: Sorting/BubbleSort.py
#Time Complexity Worst Case o(n*n-1)-- o(n^2) 
# BestCase o(n) & o(1) swaps

#Space complexity o(1)



#For first iteration n-1 comparisons and swaps

#For the second iteration n-2 comparisons and swaps.  

def bubbleSort(arr): 
    n = len(arr) 
    for i in range(n): 
        # Last i elements are already in place 
        for j in range(0, n-i-1): 
            # traverse the array from 0 to n-i-1 
            # Swap if the element found is greater 
            # than the next element 
            if arr[j] > arr[j+1] : 
                arr[j], arr[j+1] = arr[j+1], arr[j] 
  
# Driver code to test above 
arr = [60, 30, 20, 10, 22, 11, 90] 
  
bubbleSort(arr) 
  
print ("Sorted array is:") 
for i in range(len(arr)): 
    print ("%d" %arr[i]),  

File: Sorting/BubbleSort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

/*
Time Complexity
Worst Case- O(n*n)
Best Case- O(n) – When the array is already sorted
Space Complexity
Space Complexity of insertion sort is O(1)


 */


func main() {

	fmt.Println("Bubble Sort Demo ")
	rand.Seed(time.Now().UnixNano())
	min := 0
	max := 100
	i := 0
	lstIntegers := make([]int, 0)
	for i <= 10 {
		lstIntegers = append(lstIntegers, rand.Intn(max-min+1)+min)
		i = i + 1
	}
	fmt.Println("Created Random list of Integers ", lstIntegers)
	fmt.Println("Sorting it by Bubble sort")
	fmt.Println("Final Sorted List ", bubbleSort(lstIntegers))
}

func bubbleSort(input []int) []int{
    // n is the number of items in our list
    n := len(input)
    // set swapped to true
    swapped := true
    // loop
    for swapped {
        // set swapped to false
        swapped = false
        // iterate through all of the elements in our list
        for i := 1; i < n; i++ {
            // if the current element is greater than the next
            // element, swap them
            if input[i-1] > input[i] {
                // log that we are swapping values for posterity
                //fmt.Println("Swapping")
                // swap values using Go's tuple assignment
                input[i], input[i-1] = input[i-1], input[i]
                // set swapped to true - this is important
                // if the loop ends and swapped is still equal
                // to false, our algorithm will assume the list is
                // fully sorted.
                swapped = true
            }
        }
    }
    // finally, print out the sorted list
    return input
}
